Mathlib/SetTheory/Ordinal/CantorNormalForm.lean

```diff
@@ -40,45 +41,65 @@ namespace Ordinal
 /-- Inducts on the base `b` expansion of an ordinal. -/
 @[elab_as_elim]
 noncomputable def CNFRec (b : Ordinal) {C : Ordinal → Sort*} (H0 : C 0)
-    (H : ∀ o, o ≠ 0 → C (o % b ^ log b o) → C o) : ∀ o, C o := fun o ↦ by
-    by_cases h : o = 0
-    · rw [h]; exact H0
-    · exact H o h (CNFRec _ H0 H (o % b ^ log b o))
-    termination_by o => o
-    decreasing_by exact mod_opow_log_lt_self b h
+    (H : ∀ o, o ≠ 0 → C (o % b ^ log b o) → C o) (o : Ordinal) : C o :=
+  if h : o = 0 then h ▸ H0 else H o h (CNFRec _ H0 H (o % b ^ log b o))
+termination_by o
+decreasing_by exact mod_opow_log_lt_self b h
 
 @[simp]
 theorem CNFRec_zero {C : Ordinal → Sort*} (b : Ordinal) (H0 : C 0)
-    (H : ∀ o, o ≠ 0 → C (o % b ^ log b o) → C o) : @CNFRec b C H0 H 0 = H0 := by
+    (H : ∀ o, o ≠ 0 → C (o % b ^ log b o) → C o) : CNFRec b H0 H 0 = H0 := by
   rw [CNFRec, dif_pos rfl]
-  rfl
 
 theorem CNFRec_pos (b : Ordinal) {o : Ordinal} {C : Ordinal → Sort*} (ho : o ≠ 0) (H0 : C 0)
     (H : ∀ o, o ≠ 0 → C (o % b ^ log b o) → C o) :
-    @CNFRec b C H0 H o = H o ho (@CNFRec b C H0 H _) := by rw [CNFRec, dif_neg ho]
+    CNFRec b H0 H o = H o ho (@CNFRec b C H0 H _) := by
+  rw [CNFRec, dif_neg ho]
 
 /-- The Cantor normal form of an ordinal `o` is the list of coefficients and exponents in the
 base-`b` expansion of `o`.
 
-We special-case `CNF 0 o = CNF 1 o = [(0, o)]` for `o ≠ 0`.
+We special-case `CNF 0 o = CNF 1 o = [⟨0, o⟩]` for `o ≠ 0`.
 
-`CNF b (b ^ u₁ * v₁ + b ^ u₂ * v₂) = [(u₁, v₁), (u₂, v₂)]` -/
+`CNF b (b ^ u₁ * v₁ + b ^ u₂ * v₂) = [⟨u₁, v₁⟩, ⟨u₂, v₂⟩]` -/
 @[pp_nodot]
-def CNF (b o : Ordinal) : List (Ordinal × Ordinal) :=
-  CNFRec b [] (fun o _ho IH ↦ (log b o, o / b ^ log b o)::IH) o
+def CNF (b o : Ordinal) : List (Σ _ : Ordinal, Ordinal) :=
```

Comment:
I don't understand the reason for making this a Sigma type, seeing as it's not dependent. Is it only to be able to use List.keys?

Reply:
It's partially because of that, but it's also so that I can use [`AList.lookupFinsupp`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Finsupp/AList.html#AList.lookupFinsupp) to define the finitely supported function returning the CNF's coefficients in a subsequent PR (`AList` requires a Sigma type).

In practice, all this change really entails is that we write pairs as `⟨a, b⟩` instead of `(a, b)`.

Reply:
Perhaps `CNF` could directly return a `Ordinal →₀ Ordinal`, then?

Reply:
I think both forms are useful. `CNF_foldr` would be much more annoying to state with a finitely supported function.

Reply:
Wouldn't it be fairly simple with `Finset.sort`? Although maybe with addition not being commutative it does make sense to keep the order

Reply:
Hm, I suppose we could just use `Finset.sort` for that one.

Here's another example of a theorem that would be more annoying to state: `CNF (b ^ e * x + y) = CNF (b ^ e * x) ++ CNF y` for `y < b ^ e`. You could restate this with finsupps as saying the exponents of `CNF (b ^ e * x)` are all larger than those in `CNF y`, and that the coefficients of `CNF (b ^ e * x + y)` match either one of them depending if the exponent is larger or smaller than `e`, but that's a mouthful.

So yeah, I overall wouldn't want to see our list version of `CNF` go. Even if it's not the most natural for proving theorems, it makes the development of said API smoother.

Reply:
A refactor I would agree with is renaming `CNF` to something like `CNF_List`, to then use the name `CNF` for the finsupp version.

Reply:
I agree that this is not a good change, I think it would be better to just make sure you have the list functions you need

Reply:
What do you mean by that?

Reply:
What goes wrong if you use a plain product? Fix it so it doesn't go wrong anymore

Reply:
Nothing in this PR goes wrong. But it means I can't as readily use [`AList.lookupFinsupp`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Finsupp/AList.html#AList.lookupFinsupp) to define the finsupp version of the CNF in a subsequent PR.

That being said, I guess I could just use `List.map` to cast `(x, y)` into `⟨x, y⟩` in that definition. Since this change seems to be pretty controversial I think I'll just do that.

Reply:
I've been thinking about the [Veblen Normal Form](https://en.wikipedia.org/wiki/Veblen_function#Fundamental_sequences_for_the_Veblen_hierarchy) recently. That one pretty clearly needs to be a list, and if we ever want to have it in Mathlib, the easiest way is probably going to be by defining it in terms of the list form of the CNF.